lib/api/include/irods/ticketAPI.hpp

```diff
@@ -0,0 +1,52 @@
+#ifndef TICKET_API_HPP
+#define TICKET_API_HPP
+
+struct RcComm;
+
+namespace irods::ticket::administration{
+
+    int createReadTicket(RcComm* conn, char* objPath, char* ticketName);
+    int createReadTicket(RcComm* conn, char* objPath);
+
+    int createWriteTicket(RcComm* conn, char* objPath, char* ticketName);
+    int createWriteTicket(RcComm* conn, char* objPath);
+
+    int removeUsageRestriction(RcComm* conn, char* ticketName);
+    int removeUsageRestriction(RcComm* conn, int ticketID);
+
+    int removeWriteFileRestriction(RcComm* conn, char* ticketName);
+    int removeWriteFileRestriction(RcComm* conn, int ticketID);
+
+    int removeWriteByteRestriction(RcComm* conn, char* ticketName);
+    int removeWriteByteRestriction(RcComm* conn, int ticketID);
+
+    int setUsageRestriction(RcComm* conn, char* ticketName, int numUses);
+    int setUsageRestriction(RcComm* conn, int ticketID, int numUses);
+
+    int setWriteFileRestriction(RcComm* conn, char* ticketName, int numUses);
+    int setWriteFileRestriction(RcComm* conn, int ticketID, int numUses);
+
+    int setWriteByteRestriction(RcComm* conn, char* ticketName, int numUses);
+    int setWriteByteRestriction(RcComm* conn, int ticketID, int numUses);
+
+    int addUser(RcComm* conn, char* ticketName, char* user);
+    int addUser(RcComm* conn, int ticketID, char* user);
+
+    int removeUser(RcComm* conn, char* ticketName, char* user);
+    int removeUser(RcComm* conn, int ticketID, char* user);
+
+    int addGroup(RcComm* conn, char* ticketName, char* group);
+    int addGroup(RcComm* conn, int ticketID, char* group);
+
+    int removeGroup(RcComm* conn, char* ticketName, char* group);
+    int removeGroup(RcComm* conn, int ticketID, char* group);
+
+    int addHost(RcComm* conn, char* ticketName, char* host);
+    int addHost(RcComm* conn, int ticketID, char* host);
+
+    int removeHost(RcComm* conn, char* ticketName, char* host);
+    int removeHost(RcComm* conn, int ticketID, char* host);
```

Comment:
These functions are about managing constraints on a ticket. While having explicit functions for every type of constraint works, are there benefits to having a few functions that work based on data types?

For example ...
```cpp
void add_ticket_constraint(RxComm& _comm, const user_constraint& _constraint);
void remove_ticket_constraint(RxComm& _comm, const user_constraint& _constraint);

void add_ticket_constraint(RxComm& _comm, const group_constraint& _constraint);
void remove_ticket_constraint(RxComm& _comm, const group_constraint& _constraint);
```

I'm not saying your design is wrong. I'm curious to know if you have thought about this design choice.

Reply:
I'm working on this aspect currently. I'm going through different iterations to see which one would be most convenient.